### ISSUE TITLE
Trouble Installing

### ISSUE THREAD

**Issue by ellioseven** [stevenremot/ede-php-autoload#13]
Forgive me if this is a silly question, but I am trying to install this package and I cannot see it anywhere on (M)elpa. I have cloned the project and tried to load it like so:

```
(setq debug-on-error t)
(add-to-list 'load-path "~/.emacs.d/packages/ede-php-autoload/")
(load "ede-php-autoload")
```

However I am receiving an error when entering `dired-mode` and selecting a directory/file.

I receive the following log:

```
Debugger entered--Lisp error: (wrong-type-argument stringp nil)
  string-match("^/tmp_mnt/" nil)
  abbreviate-file-name(nil)
  locate-dominating-file(nil "composer.json")
  (let ((dominating-file (locate-dominating-file (buffer-file-name) ede-php-autoload-composer-file))) (if dominating-file (progn (file-name-directory dominating-file))))
  ede-php-autoload-proj-root()
  funcall(ede-php-autoload-proj-root)
  ede-project-autoload([object ede-project-autoload "php-autoload" "PHP AUTOLOAD" ede-php-autoload "composer.json" "" ede-php-autoload-proj-root nil ede-php-autoload-load ede-php-autoload-project nil nil t] "/home/elliot/Development/")
  apply(ede-project-autoload ([object ede-project-autoload "php-autoload" "PHP AUTOLOAD" ede-php-autoload "composer.json" "" ede-php-autoload-proj-root nil ede-php-autoload-load ede-php-autoload-project nil nil t] "/home/elliot/Development/"))
  eieio-generic-call-primary-only(ede-project-root-directory ([object ede-project-autoload "php-autoload" "PHP AUTOLOAD" ede-php-autoload "composer.json" "" ede-php-autoload-proj-root nil ede-php-autoload-load ede-php-autoload-project nil nil t] "/home/elliot/Development/"))
  ede-project-root-directory([object ede-project-autoload "php-autoload" "PHP AUTOLOAD" ede-php-autoload "composer.json" "" ede-php-autoload-proj-root nil ede-php-autoload-load ede-php-autoload-project nil nil t] "/home/elliot/Development/")
  #[(this dir) "\306!\307\n   \"\310\n\311\"\310\n\312\"\f;\203#
```

This makes me think I am doing something wrong here? Are there any instructions or documentation on installing and loading this package?

**Comment by stevenremot** [comment]
Hi,

The package is currently not registered on melpa! I was kind of waiting for someone to need it, so I will register it in the week ☺.

The error you got looks like a bug. If I understand correctly the first buffer you opened in your project is a dired one? In this case, would you mind trying to visit a text/code file in your project from a freshly opened emacs?

**Comment by stevenremot** [comment]
Hi,

The package is currently not registered on melpa! I was kind of waiting for someone to need it, so I will register it in the week ☺.

The error you got looks like a bug. If I understand correctly the first buffer you opened in your project is a dired one? In this case, would you mind trying to visit a text/code file in your project from a freshly opened emacs?

**Comment by ellioseven** [comment]
Thats right, I initially entered dired mode and entered a directory and I got that message. It does work however when I open up a file in a project.

**Comment by stevenremot** [comment]
Okay, the way the project tries to detect the current file/directory may not work with dired. I will have a look at it.

**Comment by stevenremot** [comment]
A quick recap of my debugging session: I tried to reproduce the error, but I don't get any EDE function called when I'm visiting a project using dired, so I could not reproduce it for now... I have to check what in my configuration prevents ede autoload to be triggered in dired buffers, this may take a bit more time than expected.

I will register the package on MELPA once the bug is fixed.

**Comment by stevenremot** [comment]
I could fix the bug with commit 26d556519599206c5e820f64cd346f714b862ba4. Can you pull master and confirm it solves your problem?

**Comment by stevenremot** [comment]
And now the package is available on melpa and melpa stable! Tell me if you have problems installing it!

**Comment by ellioseven** [comment]
Forgive me for only seeing this now! I will give this a go when I get home tonight :)

**Comment by ellioseven** [comment]
I am now receiving the following when trying to create a new php project:

```
Debugger entered--Lisp error: (args-out-of-range "" -1)
  ede-php-autoload--format-composer-pair((## . "src/") nil psr-4)
  #[257 "\302^A\300\301@#\207" [nil (psr-4 (## . "src/")) ede-php-autoload--format-composer-pair] 5 "\n\n(fn PAIR)"]((## . "src/"))
  mapcar(#[257 "\302^A\300\301@#\207" [nil (psr-4 (## . "src/")) ede-php-autoload--format-composer-pair] 5 "\n\n(fn PAIR)"] ((## . "src/")))
  ede-php-autoload--merge-composer-autoloads(((extra (branch-alias (dev-master . "3.1-dev")) (incenteev-parameters (file . "app/config/parameters.yml")) (symfony-assets-install . "relative"$
  ede-php-autoload--append-composer-autoload-data("/var/www/docroot/" nil)
  ede-php-autoload-project([object ede-php-autoload-project "PHP Autoload: /var/www/docroot/" nil "Untitled" "1.0" unbound unbound unbound nil nil unbound unbound unbound "" "" "" "" "" "" $
  apply(ede-php-autoload-project ([object ede-php-autoload-project "PHP Autoload: /var/www/docroot/" nil "Untitled" "1.0" unbound unbound unbound nil nil unbound unbound unbound "" "" "" ""$
  eieio-generic-call(initialize-instance ([object ede-php-autoload-project "PHP Autoload: /var/www/docroot/" nil "Untitled" "1.0" unbound unbound unbound nil nil unbound unbound unbound "" $
  initialize-instance([object ede-php-autoload-project "PHP Autoload: /var/www/docroot/" nil "Untitled" "1.0" unbound unbound unbound nil nil unbound unbound unbound "" "" "" "" "" "" ("deb$
  eieio-default-superclass(ede-php-autoload-project "PHP Autoload: /var/www/docroot/" :name "PHP Autoload: /var/www/docroot/" :directory "/var/www/docroot/" :file "/var/www/docroot/composer$
  apply(eieio-default-superclass (ede-php-autoload-project "PHP Autoload: /var/www/docroot/" :name "PHP Autoload: /var/www/docroot/" :directory "/var/www/docroot/" :file "/var/www/docroot/c$
  eieio-generic-call(constructor (ede-php-autoload-project "PHP Autoload: /var/www/docroot/" :name "PHP Autoload: /var/www/docroot/" :directory "/var/www/docroot/" :file "/var/www/docroot/c$
  constructor(ede-php-autoload-project "PHP Autoload: /var/www/docroot/" :name "PHP Autoload: /var/www/docroot/" :directory "/var/www/docroot/" :file "/var/www/docroot/composer.json")
  apply(constructor ede-php-autoload-project "PHP Autoload: /var/www/docroot/" (:name "PHP Autoload: /var/www/docroot/" :directory "/var/www/docroot/" :file "/var/www/docroot/composer.json"$
  ede-php-autoload-project("PHP Autoload: /var/www/docroot/" :name "PHP Autoload: /var/www/docroot/" :directory "/var/www/docroot/" :file "/var/www/docroot/composer.json")
  ede-php-autoload-load("/var/www/docroot/")
  #[(this dir) "\303^H\304\"\204^Q^@\305        !\204^Q^@\306\307!\210\303^H\310\"      !\211^Z\204 ^@\306\311!\210\312\n!)\207" [this dir o eieio-oref :safe-p ede-directory-safe-p error "A$
  apply(#[(this dir) "\303^H\304\"\204^Q^@\305  !\204^Q^@\306\307!\210\303^H\310\"      !\211^Z\204 ^@\306\311!\210\312\n!)\207" [this dir o eieio-oref :safe-p ede-directory-safe-p error "A$
  ede-auto-load-project([object ede-project-autoload "php-autoload" "PHP AUTOLOAD" ede-php-autoload "composer.json" "" ede-php-autoload-proj-root nil ede-php-autoload-load ede-php-autoload-$
  ede-load-project-file("/var/www/docroot/")
  ede("/var/www/docroot/")
  call-interactively(ede record nil)
  command-execute(ede record)
  #[257 "\304\305!\203\f^@\306\307^B!!\210^H^Y\307^A!\211^R^S\310\307^B!\311\")\207" [current-prefix-arg prefix-arg this-command real-this-command featurep smex smex-rank intern command-exe$
  ivy-call()
  ivy-read("M-x " [## ede-rescan-toplevel dired-view-file cfd-face-checklist c-digit ede-make-check-version hilit-yank debugger-previous-window sourcepattern dired-touch-program dired-faces$
  counsel-M-x()
  call-interactively(counsel-M-x nil nil)
  command-execute(counsel-M-x)
```

I used the package on MELPA with the following configuration:

```
(setq debug-on-error t)
(require 'ede)
(global-ede-mode t)

(add-to-list 'load-path "~/.emacs.d/packages/semantic-php")
(load "~/.emacs.d/packages/semantic-php/loaddefs.el")
(add-hook 'php-mode-hook #'semantic-mode)

(use-package ede-php-autoload
  :ensure t)
```

**Comment by stevenremot** [comment]
Duh, I hope you would make me happy this time :'(

It seems there have been en error when parsing the composer.json file.
Would you mind showing it so that I can test?

Le dim. 9 oct. 2016 06:38, Elliot Mitchum <notifications@github.com> a
écrit :

**Comment by ellioseven** [comment]
Hopefully it's just me being silly and doing it wrong:

```
{
    "name": "symfony/framework-standard-edition",
    "license": "MIT",
    "type": "project",
    "description": "The \"Symfony Standard Edition\" distribution",
    "autoload": {
        "psr-4": { "": "src/" },
        "classmap": [ "app/AppKernel.php", "app/AppCache.php" ]
    },
    "autoload-dev": {
        "psr-4": { "Tests\\": "tests/" }
    },
    "require": {
        "php": ">=5.5.9",
        "symfony/symfony": "3.1.*",
        "doctrine/orm": "^2.5",
        "doctrine/doctrine-bundle": "^1.6",
        "doctrine/doctrine-cache-bundle": "^1.2",
        "symfony/swiftmailer-bundle": "^2.3",
        "symfony/monolog-bundle": "^2.8",
        "symfony/polyfill-apcu": "^1.0",
        "sensio/distribution-bundle": "^5.0",
        "sensio/framework-extra-bundle": "^3.0.2",
        "incenteev/composer-parameter-handler": "^2.0"
    },
    "require-dev": {
        "sensio/generator-bundle": "^3.0",
        "symfony/phpunit-bridge": "^3.0"
    },
    "scripts": {
        "symfony-scripts": [
            "Incenteev\\ParameterHandler\\ScriptHandler::buildParameters",
            "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::buildBootstrap",
            "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::clearCache",
            "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::installAssets",
            "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::installRequirementsFile",
            "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::prepareDeploymentTarget"
        ],
        "post-install-cmd": [
            "@symfony-scripts"
        ],
        "post-update-cmd": [
            "@symfony-scripts"
        ]
    },
    "config": {
        "platform": {
            "php": "5.5.9"
        }
    },
    "extra": {
        "symfony-app-dir": "app",
        "symfony-bin-dir": "bin",
        "symfony-var-dir": "var",
        "symfony-web-dir": "web",
        "symfony-tests-dir": "tests",
        "symfony-assets-install": "relative",
        "incenteev-parameters": {
            "file": "app/config/parameters.yml"
        },
        "branch-alias": {
            "dev-master": "3.1-dev"
        }
    }
}
```

**Comment by stevenremot** [comment]
Aha, the fallback directory line ` "": "src/"` isn't supported by ede-php-autoload for now. I wasn't aware of this feature. I will try to come back to you with a proper implementation of it in the next days.

**Comment by ellioseven** [comment]
Haha what are the odds!

**Comment by stevenremot** [comment]
Yeah, we will end up fixing all ed-php-autoload problems in one single issue!

**Comment by stevenremot** [comment]
Fixed in tag 0.4.2! Does it work on your side?

**Comment by stevenremot** [comment]
Hey @ellioseven! Do you mind if I close this issue? I suppose it works if no one came back to complain!
